Compress ValueChunk with zstd

metalmatze opened the pull request:
Compress all samples in a valueChunk with zstd.
All samples are appended to the chunk without any compression (needs to be updated in conprof itself) and whenever `valuesChunk.Bytes()` is called it'll compress all samples in the chunk with zstd before returning the compressed `[]byte`.
Similarly when iterating over samples of a chunk the iterator checks if the chunk is available decompressed and if not then we'll decompress with zstd "on-the-fly" and write the decompressed contents to `c.b` dropping `c.compressed`.

Benchmarks have shown that the compression doesn't significantly improve after having 1000 samples in each chunk. This however, is way to much in terms of in-memory consumption and we'll probably should stay around 10-20 samples per chunk.
Nonetheless I've seen 50% disk savings with 20 samples per chunk.

![Screenshot from 2021-03-03 12-53-58](https://user-images.githubusercontent.com/872251/109830926-3fcb3d00-7c3f-11eb-8d2f-d09a80343d6e.png)
_Compares conprof-dev with 20 samples per chunk with zstd to conprof-master with 12 samples per chunk individually compressed with gzip._

metalmatze committed:
Compress chunks with zstd when Bytes() is called

Especially if the samples are uncompressed the real benefit comes from
compressing 100-1000 samples at once.

metalmatze committed:
Compress valueChunk with zstd in Bytes() and decompress in Iterator

metalmatze committed:
Move benchmark with specific dataset behind build tag

These benchmarks need some sort of dataset. It doesn't need to be a very
specific one, but 100+ pb.gz files should do it for a start.

metalmatze committed:
Fix linting and update testdata/index_format_v1

metalmatze committed:
Defer closing zstd decoder in valueChunk Iterator

thorfour commented on tsdb/chunkenc/byte_test.go:
```diff
@@ -9,13 +9,19 @@ import (
 
 func TestLoadBytesChunk(t *testing.T) {
-	// tree samples added (0,conprof) (1,conprof) (2,conprof)
+	//c := NewBytesChunk()
```
forgot to remove?

metalmatze replied:
This is how the static `byte{}` was generated. I can add a comment if you want me to. This is just so that we can check if `LoadBytesChunk()` works correctly.

thorfour commented on tsdb/chunkenc/value.go:
```diff
@@ -16,23 +16,44 @@ package chunkenc
+	"github.com/klauspost/compress/zstd"
 )
 
+// zstdFrameMagic is the magic beginning of all zstd compressed frames.
```
is this a zstd standard? If so, can you link the spec?

metalmatze replied:
Yes, it's a zstd standard and copied from the compress zstd lib we're using.
https://github.com/klauspost/compress/blob/063ee1dad7a10b7caef0432d813ec3b8d72c8a8f/zstd/framedec.go#L59
https://www.rfc-editor.org/rfc/rfc8478.html#section-3.1.1

metalmatze replied:
I guess I'll add this to the source itself :)

thorfour replied:
Thank you!

thorfour commented on tsdb/chunkenc/value.go:
```diff
@@ -16,23 +16,44 @@ package chunkenc
+	compressed := &bytes.Buffer{}
+	encoder, err := zstd.NewWriter(compressed, zstd.WithEncoderLevel(zstd.SpeedFastest))
+	if err != nil {
+		panic(err) // TODO don't panic
```
did you mean to leave these TODO's in here?

metalmatze replied:
Yeah, I not sure myself what to do in these cases.
I can remove the comments and leave the panics, or we ignore the errors...
Either way not perfect.

thorfour replied:
I'm never a fan of panics in code. Do we have a logger that we could log the errors here?

thorfour replied:
Log the error and return nil?

metalmatze replied:
Nope. Nothing at most `log.Println` from stdlib...

metalmatze replied:
I guess the question boils down to. Do we panic and stop the entire system if we cannot decompress one valueChunk or do we simply not care and return nil?
I'm probably leaning to the latter and that would mean ignoring errors for the most part.

brancz replied:
This is usually a case for an error iterator.

metalmatze replied:
Good idea. I create the valueIterator first and the set the error on that iterator in case it happens during the decompression. That way by the time `Next()` is called it returns `false` and the error gets surfaced through `it.Err()`.

thorfour reviewed (commented):
Awesome! Couple of small questions

metalmatze commented:
For completeness I've deployed conprof-dev with 12 chunks at 15:30 too. As we can see it's very close to the original one, while saving disk.

![Screenshot from 2021-03-03 19-01-58](https://user-images.githubusercontent.com/872251/109850879-4b286380-7c53-11eb-8bd3-bddebb63222c.png)

brancz commented:
In that case let's go with 12 for now? I feel like there is still a lot to be gained here but this is a good start.

metalmatze commented:
Yes, let's keep it at 12 to minimize the difference for now. We can still increase later on.
I might try and see how the interface look if we were to change `Bytes() []byte` to `Bytes() ([]byte, error)` tomorrow.

metalmatze committed:
Chunk interface Bytes() now returns ([]byte, error)

This is needed as in Bytes() we now compress the chunk's data with zstd
and it could fail compressing. Rather than panicing or ignoring the
error we try to bubble it up to the caller.

metalmatze committed:
Surface valueIterator decompression errors through iterator error

metalmatze committed:
Add comment on where zstdFrameMagic is coming from

brancz reviewed (approved):
lgtm but giving @thorfour another chance to review, nice work!

metalmatze committed:
Store compressed valueChunk bytes in chunk in case it gets read again

If the len(compressed) != 0 while appending a new sample we'll
invalidate the cached compressed data and will need to compress again
whenever Bytes() is called.

metalmatze committed:
Add comment about how to generate bytes for test

metalmatze committed:
Return after setting an error in valueChunk.Iterator

metalmatze committed:
Trigger GitHub Actions

For some reason the CI didn't run...

metalmatze committed:
Don't compress empty valueChunk

thorfour reviewed (approved):
Looks great!
